rclcpp_generic/include/rclcpp_generic/generic_subscription.hpp

```diff
@@ -0,0 +1,115 @@
+{
+
+/**
+ * This class is an implementation of an rclcpp::SubscriptionBase for serialized messages whose topic
+ * is not known at compile time (hence templating does not work).
+ *
+ * It does not support intra-process handling.
+ */
+class GenericSubscription : public rclcpp::SubscriptionBase
+{
+public:
+  RCLCPP_SMART_PTR_DEFINITIONS(GenericSubscription)
+
+  /**
+   * Factory method.
+   *
+   * The returned pointer will never be empty, but this function can throw various exceptions, for
+   * instance when the message's package can not be found on the AMENT_PREFIX_PATH.
+   *
+   * \param topics_interface NodeTopicsInterface pointer used in parts of the setup.
+   * \param topic_name Topic name
+   * \param topic_type Topic type
+   * \param qos QoS settings
+   * \param callback Callback for new messages of serialized form
+   * \param group Callback group
+   */
+  static std::shared_ptr<GenericSubscription> create(
+    rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr topics_interface,
+    const std::string & topic_name,
+    const std::string & topic_type,
+    const rclcpp::QoS & qos,
+    std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)> callback,
+    rclcpp::CallbackGroup::SharedPtr group = nullptr);
+
+  /**
+   * Constructor. In order to properly subscribe to a topic, this subscription needs to be added to
+   * the node_topic_interface of the node passed into this constructor.
+   *
+   * \param node_base NodeBaseInterface pointer used in parts of the setup.
+   * \param ts_lib Type support library for this type
+   * \param topic_name Topic name
+   * \param topic_type Topic type
+   * \param qos QoS settings
+   * \param callback Callback for new messages of serialized form
+   */
+  GenericSubscription(
+    rclcpp::node_interfaces::NodeBaseInterface * node_base,
+    const std::shared_ptr<rcpputils::SharedLibrary> ts_lib,
+    const std::string & topic_name,
+    const std::string & topic_type,
+    const rclcpp::QoS & qos,
+    std::function<void(std::shared_ptr<rclcpp::SerializedMessage>)> callback);
+
+  // Same as create_serialized_message() as the subscription is to serialized_messages only
+  std::shared_ptr<void> create_message() override;
+
+  std::shared_ptr<rclcpp::SerializedMessage> create_serialized_message() override;
+
+  void handle_message(
+    std::shared_ptr<void> & message, const rclcpp::MessageInfo & message_info) override;
+
+  void handle_loaned_message(
+    void * loaned_message, const rclcpp::MessageInfo & message_info) override;
+
+  // Same as return_serialized_message() as the subscription is to serialized_messages only
+  void return_message(std::shared_ptr<void> & message) override;
+
+  void return_serialized_message(std::shared_ptr<rclcpp::SerializedMessage> & message) override;
+
+  // Provide a const reference to the QoS Profile used to create this subscription.
+  const rclcpp::QoS & qos_profile() const;
+
+private:
+  RCLCPP_DISABLE_COPY(GenericSubscription)
```

Comment:
Copied from `rosbag2` – does copying really need to be disabled, and if yes, does that apply to the publisher as well?

Reply:
I think we disable copying on all publishers and subscriptions because they have unique RMW implemntation resources under the hood - so it can cause nondeterministic behavior if those resources are shared: e.g. one copy will get the first message, then the other copy will get the next message, but neither will get all messages, they'll consume in some order.

Reply:
Shouldn't the publisher and subscriber base classes then disable copying?

Reply:
Bump – I'm unsure if the publisher also needs to disable copying, or if I should even add it to the base classes.